buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -545,11 +560,74 @@ static int writeUtf8(AbstractByteBuf buffer, int writerIndex, CharSequence seq,
         return writerIndex - oldWriterIndex;
     }
 
+    /**
+     * Returns max bytes length of UTF8 character sequence of the given length.
+     */
+    public static int utf8MaxBytes(final int seqLength) {
+        return seqLength * MAX_BYTES_PER_CHAR_UTF8;
+    }
+
     /**
      * Returns max bytes length of UTF8 character sequence.
+     * <p>
+     * It behaves like {@link #utf8MaxBytes(int)} applied to {@code seq} {@link CharSequence#length()}.
      */
     public static int utf8MaxBytes(CharSequence seq) {
-        return seq.length() * MAX_BYTES_PER_CHAR_UTF8;
+        return utf8MaxBytes(seq.length());
+    }
+
+    /**
+     * Returns the exact bytes length of UTF8 character sequence.
+     * <p>
+     * This method is producing the exact length according to {@link #writeUtf8(ByteBuf, CharSequence)}.
+     */
+    public static int utf8Bytes(final CharSequence seq) {
+        int seqLength = seq.length();
+        int i = 0;
+        // ASCII fast path
+        while (i < seqLength && seq.charAt(i) < 0x80) {
+            ++i;
```

Comment:
If you are in the nitty gritty of optimizing this, consider iterating in reverse.  Checking larger indicies before smaller ones allows bounds check elimination, and usually the cache prefetcher does work backwards.  

If not, you can ignore.

Reply:
Re bounds checking I've just verified that with ascii cases it already works as expected, otherwise it won't make any difference due to `InlineSmallCode` that prevent the uncommon path from being inlined and forcing the JVM to put bounds check before the uncommon loop too.